### PR TITLE
[UBS] - Tests should be written for the deactivateTariffForChosenParam() method in the class SuperAdminServiceImpl.class #4855

### PR DESCRIPTION
**Pull request by juseti**
Tests were written for the deactivateTariffForChosenParam method in the SuperAdminServiceImpl class (https://github.com/ita-social-projects/GreenCity/issues/4855) that meet the requirements https://github.com/ita-social-projects/GreenCity/issues/4762

1. Both ordinary and parameterized tests were written for different parameter cases.
2. Added a method to detect unverified interactions on mock objects after each test run.
3. Added missing verifications for mocked objects.
4. Methods to get DetailsOfDeactivateTariffsDto objects with different sets of parameters are implemented in the ModelUtils class.

### CODE REVIEW

File: service/src/test/java/greencity/ModelUtils.java [hunk 1]
```diff
@@ -7,11 +7,7 @@
 import com.google.maps.model.Geometry;
 import com.google.maps.model.LatLng;
 import greencity.constant.AppConstant;
-import greencity.dto.AddNewTariffDto;
-import greencity.dto.CreateAddressRequestDto;
-import greencity.dto.LocationsDtos;
-import greencity.dto.OptionForColumnDTO;
-import greencity.dto.TariffsForLocationDto;
+import greencity.dto.*;
```
Comment (RomaMocherniuk) [inline service/src/test/java/greencity/ModelUtils.java]: please do not use `*`

File: service/src/test/java/greencity/service/ubs/SuperAdminServiceImplTest.java [hunk 2]
```diff
@@ -859,4 +868,294 @@ void changeTariffLocationsStatusParamUnresolvable() {
...
+        verify(deactivateTariffsForChosenParamRepository).deactivateTariffsByRegions(List.of(1L, 2L));
+    }
+
+    @Test
+    void deactivateTariffByNotExistingRegionThrows() {
+        DetailsOfDeactivateTariffsDto details = ModelUtils.getDetailsOfDeactivateTariffsDtoWithRegion();
+
+        when(deactivateTariffsForChosenParamRepository.isRegionsExists(anyList())).thenReturn(false);
+        assertThrows(NotFoundException.class,
+            () -> superAdminService.deactivateTariffForChosenParam(details));
```
Comment (RomaMocherniuk) [inline service/src/test/java/greencity/service/ubs/SuperAdminServiceImplTest.java:941]: please add `verify`

File: service/src/test/java/greencity/service/ubs/SuperAdminServiceImplTest.java [hunk 3]
```diff
@@ -859,4 +868,294 @@ void changeTariffLocationsStatusParamUnresolvable() {
...
+    }
+
+    @Test
+    void deactivateTariffByOneRegionAndCities() {
+        DetailsOfDeactivateTariffsDto details = ModelUtils.getDetailsOfDeactivateTariffsDtoWithRegionAndCities();
+
+        when(regionRepository.existsRegionById(anyLong())).thenReturn(true);
+        when(deactivateTariffsForChosenParamRepository.isCitiesExistForRegion(anyList(), anyLong())).thenReturn(true);
+        superAdminService.deactivateTariffForChosenParam(details);
+        verify(deactivateTariffsForChosenParamRepository).deactivateTariffsByRegionsAndCities(List.of(1L, 11L), 1L);
```
Comment (RomaMocherniuk) [inline service/src/test/java/greencity/service/ubs/SuperAdminServiceImplTest.java:955]: please add `verify` for all mooks

File: service/src/test/java/greencity/service/ubs/SuperAdminServiceImplTest.java [hunk 4]
```diff
@@ -859,4 +868,294 @@ void changeTariffLocationsStatusParamUnresolvable() {
         assertThrows(BadRequestException.class,
             () -> superAdminService.changeTariffLocationsStatus(1L, dto, "unresolvable"));
     }
+
+    @Test
+    void deactivateTariffByOneRegion() {
+        DetailsOfDeactivateTariffsDto details = ModelUtils.getDetailsOfDeactivateTariffsDtoWithRegion();
+        when(deactivateTariffsForChosenParamRepository.isRegionsExists(anyList())).thenReturn(true);
+        superAdminService.deactivateTariffForChosenParam(details);
+        verify(deactivateTariffsForChosenParamRepository).deactivateTariffsByRegions(List.of(1L));
```
Comment (RomaMocherniuk) [inline service/src/test/java/greencity/service/ubs/SuperAdminServiceImplTest.java:921]: please check in every test if there are no more interactions with the mocks

File: service/src/test/java/greencity/service/ubs/SuperAdminServiceImplTest.java [hunk 5]
```diff
@@ -859,4 +868,294 @@ void changeTariffLocationsStatusParamUnresolvable() {
...
+            arguments(false, false, true, true),
+            arguments(false, true, false, true),
+            arguments(false, true, true, false),
+            arguments(true, false, false, true),
+            arguments(true, false, true, false),
+            arguments(true, true, false, false),
+            arguments(false, false, false, true),
+            arguments(false, false, true, false),
+            arguments(true, false, false, false),
+            arguments(false, false, false, false));
```
Comment (RomaMocherniuk) [inline service/src/test/java/greencity/service/ubs/SuperAdminServiceImplTest.java:1188]: Should we add 
```
    true, true, true, true	
    false, true, false, false
```
to this list?

File: service/src/test/java/greencity/service/ubs/SuperAdminServiceImplTest.java [hunk 6]
```diff
@@ -859,4 +868,294 @@ void changeTariffLocationsStatusParamUnresolvable() {
...
+            when(deactivateTariffsForChosenParamRepository
+                .isCitiesExistForRegion(anyList(), anyLong())).thenReturn(isCitiesExistForRegion);
+            if (isCitiesExistForRegion) {
+                when(deactivateTariffsForChosenParamRepository
+                    .isReceivingStationsExists(anyList())).thenReturn(isReceivingStationsExists);
+            }
+        }
+    }
+
+    private void doMockForTariffWithAllParams(boolean isRegionExists,
```
Comment (RomaMocherniuk) [inline service/src/test/java/greencity/service/ubs/SuperAdminServiceImplTest.java]: please move `boolean isRegionExists` to the new line

File: service/src/test/java/greencity/service/ubs/SuperAdminServiceImplTest.java [hunk 7]
```diff
@@ -859,4 +868,294 @@ void changeTariffLocationsStatusParamUnresolvable() {
...
+                .setRegionsId(Optional.of(List.of(1L, 2L)))),
+            arguments(ModelUtils.getDetailsOfDeactivateTariffsDtoWithRegionAndCities()
+                .setRegionsId(Optional.of(List.of(1L, 2L)))),
+            arguments(ModelUtils.getDetailsOfDeactivateTariffsDtoWithRegionAndCityAndStation()
+                .setRegionsId(Optional.of(List.of(1L, 2L)))),
+            arguments(ModelUtils.getDetailsOfDeactivateTariffsDtoWithAllParams()
+                .setRegionsId(Optional.of(List.of(1L, 2L)))));
+    }
+
+    private void doMockForTariffWithRegionAndCityAndStation(boolean isRegionExists,
```
Comment (RomaMocherniuk) [inline service/src/test/java/greencity/service/ubs/SuperAdminServiceImplTest.java]: please move `boolean isRegionExists` to the new line

File: service/src/test/java/greencity/service/ubs/SuperAdminServiceImplTest.java [hunk 8]
```diff
@@ -859,4 +868,294 @@ void changeTariffLocationsStatusParamUnresolvable() {
...
+
+        doMockForTariffWithAllParams(true, true, true, true);
+        superAdminService.deactivateTariffForChosenParam(details);
+        verify(deactivateTariffsForChosenParamRepository)
+            .deactivateTariffsByAllParam(1L, List.of(1L, 11L), List.of(1L, 12L), 1L);
+    }
+
+    @ParameterizedTest
+    @MethodSource("deactivateTariffByAllWithNotExistingParamProvider")
+    void deactivateTariffByAllWithNotExistingParamThrows(boolean isRegionExists,
```
Comment (RomaMocherniuk) [inline service/src/test/java/greencity/service/ubs/SuperAdminServiceImplTest.java]: same here

File: service/src/test/java/greencity/service/ubs/SuperAdminServiceImplTest.java (hunk 5 above)
Comment (juseti) [reply service/src/test/java/greencity/service/ubs/SuperAdminServiceImplTest.java:1188]: true, true, true, true - this case is without exception.

File: service/src/test/java/greencity/service/ubs/SuperAdminServiceImplTest.java (hunk 4 above)
Comment (juseti) [reply service/src/test/java/greencity/service/ubs/SuperAdminServiceImplTest.java:921]: checked

File: service/src/test/java/greencity/service/ubs/SuperAdminServiceImplTest.java (hunk 2 above)
Comment (juseti) [reply service/src/test/java/greencity/service/ubs/SuperAdminServiceImplTest.java:941]: added

File: service/src/test/java/greencity/service/ubs/SuperAdminServiceImplTest.java (hunk 3 above)
Comment (juseti) [reply service/src/test/java/greencity/service/ubs/SuperAdminServiceImplTest.java:955]: added

File: service/src/test/java/greencity/service/ubs/SuperAdminServiceImplTest.java (hunk 5 above)
Comment (juseti) [reply service/src/test/java/greencity/service/ubs/SuperAdminServiceImplTest.java:1188]: false, true, false, false - added

File: service/src/test/java/greencity/ModelUtils.java (hunk 1 above)
Comment (juseti) [reply service/src/test/java/greencity/ModelUtils.java]: fixed

File: service/src/test/java/greencity/service/ubs/SuperAdminServiceImplTest.java [hunk 9]
```diff
@@ -859,4 +865,382 @@ void changeTariffLocationsStatusParamUnresolvable() {
...
+    void deactivateTariffByOneRegionAndCities() {
+        DetailsOfDeactivateTariffsDto details = ModelUtils.getDetailsOfDeactivateTariffsDtoWithRegionAndCities();
+
+        when(regionRepository.existsRegionById(anyLong())).thenReturn(true);
+        when(deactivateTariffsForChosenParamRepository.isCitiesExistForRegion(anyList(), anyLong())).thenReturn(true);
+        superAdminService.deactivateTariffForChosenParam(details);
+        verify(regionRepository).existsRegionById(1L);
+        verify(deactivateTariffsForChosenParamRepository).isCitiesExistForRegion(List.of(1L, 11L), 1L);
+        verify(deactivateTariffsForChosenParamRepository).deactivateTariffsByRegionsAndCities(List.of(1L, 11L), 1L);
+        verifyNoMoreInteractions(regionRepository, deactivateTariffsForChosenParamRepository);
```
Comment (RomaMocherniuk) [inline service/src/test/java/greencity/service/ubs/SuperAdminServiceImplTest.java]: please move this verification to the `After` method and add all mocks as input params

Comment (juseti) [reply service/src/test/java/greencity/service/ubs/SuperAdminServiceImplTest.java]: done

File: service/src/test/java/greencity/service/ubs/SuperAdminServiceImplTest.java [hunk 10]
```diff
@@ -80,9 +88,26 @@ class SuperAdminServiceImplTest {
...
     @Mock
     private TariffLocationRepository tariffsLocationRepository;
-
     @Mock
-    private DeactivateChosenEntityRepository deactivateChosenEntityRepository;
+    private DeactivateChosenEntityRepository deactivateTariffsForChosenParamRepository;
+
+    @AfterEach
+    void afterEach() {
+        verifyNoMoreInteractions(
```
Comment (RomaMocherniuk) [inline service/src/test/java/greencity/service/ubs/SuperAdminServiceImplTest.java:96]: 👍

**Review by RomaMocherniuk — APPROVED** [review]

**Review by OlegVat — APPROVED** [review]

**Review by NKorzhik — APPROVED** [review]

**Review by JJJazl — APPROVED** [review]

**Pull request merged**